arch/risc-v/src/mpfs/mpfs_usb.c

```diff
@@ -0,0 +1,3770 @@
+
+  for (epno = 0, bit = 1, epset &= MPFS_EPSET_ALL;
+       epno < MPFS_USB_NENDPOINTS && epset != 0;
+       epno++, bit <<= 1)
+    {
+      /* Is this endpoint in the set? */
+
+      if ((epset & bit) != 0)
+        {
+          /* Yes, reset and disable it */
+
+          mpfs_ep_reset(priv, epno);
+          epset &= ~bit;
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: mpfs_pullup
+ *
+ * Description:
+ *   This is the pullup() method of the USB device driver interface. This is
+ *   a no-op.
+ *
+ * Input Parameters:
+ *   dev       - USB device
+ *   enable    - Enable or disable
+ *
+ * Returned Value:
+ *   OK unconditionally
+ *
+ ****************************************************************************/
+
+static int mpfs_pullup(struct usbdev_s *dev, bool enable)
+{
+  return OK;
+}
+
+/****************************************************************************
+ * Initialization / Reset
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mpfs_usb_iomux
+ *
+ * Description:
+ *   This initializes the necessary pin-muxes for the USB controller.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void mpfs_usb_iomux(void)
+{
+  putreg32(LIBERO_SETTING_IOMUX3_CR, MPFS_SYSREG_IOMUX3);
+  putreg32(LIBERO_SETTING_IOMUX4_CR, MPFS_SYSREG_IOMUX4);
+
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_CFG_CR, MPFS_SYSREG_B2_CFG);
+
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_0_1_CR, MPFS_SYSREG_B2_0_1);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_2_3_CR, MPFS_SYSREG_B2_2_3);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_4_5_CR, MPFS_SYSREG_B2_4_5);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_6_7_CR, MPFS_SYSREG_B2_6_7);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_8_9_CR, MPFS_SYSREG_B2_8_9);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_10_11_CR,
+           MPFS_SYSREG_B2_10_11);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_12_13_CR,
+           MPFS_SYSREG_B2_12_13);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_14_15_CR,
+           MPFS_SYSREG_B2_14_15);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_16_17_CR,
+           MPFS_SYSREG_B2_16_17);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_18_19_CR,
+           MPFS_SYSREG_B2_18_19);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_20_21_CR,
+           MPFS_SYSREG_B2_20_21);
+  putreg32(LIBERO_SETTING_MSSIO_BANK2_IO_CFG_22_23_CR,
+           MPFS_SYSREG_B2_22_23);
+
+#ifdef CONFIG_USBDEV_DMA
+  /* DMA operations need to open the USB PMP registers for proper
+   * operation. If not configured, apply default settings.
+   */
+
+  uint64_t pmpcfg_usb_x;
+
+  pmpcfg_usb_x = getreg64(MPFS_PMPCFG_USB_0);
+  if ((pmpcfg_usb_x & 0x1ffffff000000000) != 0x1f00000000000000)
```

Comment:
will `0x1ffffff000000000` work? I think `0x1ffffff000000000ll` is needed

Reply:
I think it does, we use the same in the emmc driver. (without the ll's)

Reply:
I think that for 64bit targets it might be ok. I will make some experiments and feedback. Anyway adding llu will not bring any harm